internal/overlord/cmdstate/cmdmgr.go

```diff
@@ -0,0 +1,654 @@
+			return err
+		}
+		ptys = append(ptys, stderrReader)
+		ttys = append(ttys, stderrWriter)
+		stderr = stderrWriter
+		stderrConn := e.getWebsocket(wsStderr)
+		wg.Add(1)
+		go func() {
+			<-wsutil.WebsocketSendStream(stderrConn, stderrReader, -1)
+			stderrReader.Close()
+			wg.Done()
+		}()
+	}
+
+	if e.timeout != 0 {
+		var cancel func()
+		ctx, cancel = context.WithTimeout(ctx, e.timeout)
+		defer cancel()
+	}
+
+	cmd := exec.CommandContext(ctx, e.command[0], e.command[1:]...)
+
+	for k, v := range e.env {
+		cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%s", k, v))
+	}
+
+	cmd.Dir = e.workingDir
+
+	cmd.Stdin = stdin
+	cmd.Stdout = stdout
+	cmd.Stderr = stderr
+
+	cmd.SysProcAttr = &syscall.SysProcAttr{}
+	if e.uid != nil && e.gid != nil {
+		cmd.SysProcAttr.Credential = &syscall.Credential{
+			Uid: uint32(*e.uid),
+			Gid: uint32(*e.gid),
+		}
+	}
+
+	// Creates a new session if the calling process is not a process group
+	// leader. The calling process is the leader of the new session, the
+	// process group leader of the new process group, and has no controlling
+	// terminal. This is important to allow remote shells to handle Ctrl+C.
+	cmd.SysProcAttr.Setsid = true
+
+	// Make the given terminal the controlling terminal of the calling
+	// process. The calling process must be a session leader and not have a
+	// controlling terminal already. This is important as allows Ctrl+C to
+	// work as expected for non-shell programs.
+	if e.useTerminal {
+		cmd.SysProcAttr.Setctty = true
+	}
+
+	finisher := func(exitCode int, cmdErr error) error {
+		for _, tty := range ttys {
+			_ = tty.Close()
+		}
+
+		// Close the control channel, if connected.
+		controlConn := e.getWebsocket(wsControl)
+		if controlConn != nil {
+			_ = controlConn.Close()
+		}
+
+		close(childDead)
+
+		wg.Wait()
+
+		for _, pty := range ptys {
+			_ = pty.Close()
+		}
+
+		setApiData(change, exitCode)
+
+		return cmdErr
+	}
+
+	// Start the command!
+	err = cmd.Start()
+	if err != nil {
+		return finisher(-1, err)
+	}
+
+	// Send its PID to the control loop.
+	pidCh <- cmd.Process.Pid
+
+	// Wait for it to finish.
+	err = cmd.Wait()
+
+	// Handle errors: timeout, non-zero exit code, or other error.
+	if errors.Is(ctx.Err(), context.DeadlineExceeded) {
+		return finisher(-1, fmt.Errorf("timed out after %v: %w", e.timeout, ctx.Err()))
+	} else if exitErr, ok := err.(*exec.ExitError); ok {
+		status, ok := exitErr.Sys().(syscall.WaitStatus)
+		if ok {
+			if status.Signaled() {
+				// 128 + n == Fatal error signal "n"
+				return finisher(128+int(status.Signal()), nil)
+			}
+			return finisher(status.ExitStatus(), nil)
+		}
+		return finisher(-1, err)
+	} else if err != nil {
+		return finisher(-1, err)
+	}
+
+	// Successful exit (exit code 0).
+	return finisher(0, nil)
+}
+
+func setApiData(change *state.Change, exitCode int) {
+	st := change.State()
+	st.Lock()
+	defer st.Unlock()
+	change.Set("api-data", map[string]interface{}{
+		"exit-code": exitCode,
```

Comment:
Another symptom of the earlier note: this is really the exit code of the task.  If it needs to be exposed, ideally it would be the piece that aggregates tasks doing it.

Reply:
Okay, I'm setting this on the task now, and then the code that aggregates the tasks into change info grabs task `api-data` if it's present, otherwise change `api-data` overrides that if that's present.